### PR TITLE
Key generation with some assumptions

### PR DESCRIPTION
**Pull request by Elfoniok**

### CODE REVIEW

File: peersession_test.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ import (
 
 const (
 	TEST_NAME     = "bootstrap-unittest"
+	KEY_DIFF      = 10
```
Comment (Krigpl) [inline peersession_test.go]: Just put `0`, it's not important in this test and it shouldn't slow it down

File: crypto/crypto.go [hunk 2]
```diff
@@ -20,8 +20,30 @@ func GetPubKeyHex(privateKey PrivateKey) string {
 	return hex.EncodeToString(privKey.PublicKey.X) + hex.EncodeToString(privKey.PublicKey.Y)
 }
 
-func GeneratePrivateKey() (PrivateKey, error) {
-	return elliptic.GeneratePrivateKey(elliptic.Secp256k1)
+func GeneratePrivateKey(difficulty int) (PrivateKey, error) {
+	for true {
+		privKey, err := elliptic.GeneratePrivateKey(elliptic.Secp256k1)
+		pubKey := sha256.Sum256(append(privKey.PublicKey.X, privKey.PublicKey.X...))
```
Comment (Krigpl) [inline crypto/crypto.go]: How about `privKey.PublicKey.X, privKey.PublicKey.Y...`? :)
Also this is incorrect
`append(a, b...)` may modify `a` which isn't desired here. You need to create an extra byte array and copy X,Y there

File: crypto/crypto.go [hunk 3]
```diff
@@ -20,8 +20,30 @@ func GetPubKeyHex(privateKey PrivateKey) string {
...
+		isDifficult := true
+
+		for i := 0; i < nullBytes; i++ {
+			if pubKey[i] != 0 {
+				isDifficult = false
+				break
+			}
+		}
+
+		if isDifficult == true {
```
Comment (Krigpl) [inline crypto/crypto.go]: just `if isDifficult {`

File: crypto/crypto.go [hunk 4]
```diff
@@ -20,8 +20,30 @@ func GetPubKeyHex(privateKey PrivateKey) string {
...
+			if pubKey[nullBytes] <= (2<<(8-uint(remainder)-1))-1 {
```
Comment (Krigpl) [inline crypto/crypto.go]: it's the same as
`if pubKey[nullBytes] < (1<<(8-uint(remainder))) {`

also you may do `uint` cast at the declaration to avoid it here to make it more readable

File: crypto/crypto.go [hunk 5]
```diff
@@ -20,8 +20,30 @@ func GetPubKeyHex(privateKey PrivateKey) string {
 	return hex.EncodeToString(privKey.PublicKey.X) + hex.EncodeToString(privKey.PublicKey.Y)
 }
 
-func GeneratePrivateKey() (PrivateKey, error) {
-	return elliptic.GeneratePrivateKey(elliptic.Secp256k1)
+func GeneratePrivateKey(difficulty int) (PrivateKey, error) {
+	for true {
+		privKey, err := elliptic.GeneratePrivateKey(elliptic.Secp256k1)
```
Comment (Krigpl) [inline crypto/crypto.go]: check for `err` and don't proceed if it's not null

File: crypto/crypto.go [hunk 6]
```diff
@@ -20,8 +20,30 @@ func GetPubKeyHex(privateKey PrivateKey) string {
...
+		for i := 0; i < nullBytes; i++ {
+			if pubKey[i] != 0 {
+				isDifficult = false
+				break
+			}
+		}
+
+		if isDifficult == true {
+			if pubKey[nullBytes] <= (2<<(8-uint(remainder)-1))-1 {
+				return privKey, err
```
Comment (Krigpl) [inline crypto/crypto.go]: `err` is null at this point

**Review by Krigpl — CHANGES_REQUESTED** [review]
also please add test :)

File: peersession_test.go (hunk 1 above)
Comment (Elfoniok) [reply peersession_test.go]: Done

File: crypto/crypto.go (hunk 2 above)
Comment (Elfoniok) [reply crypto/crypto.go]: Done

File: crypto/crypto.go (hunk 3 above)
Comment (Elfoniok) [reply crypto/crypto.go]: Done

File: crypto/crypto.go (hunk 4 above)
Comment (Elfoniok) [reply crypto/crypto.go]: Nice simplification

File: crypto/crypto.go (hunk 6 above)
Comment (Elfoniok) [reply crypto/crypto.go]: right

File: crypto/crypto.go (hunk 5 above)
Comment (Elfoniok) [reply crypto/crypto.go]: done

File: crypto/crypto.go [hunk 7]
```diff
@@ -20,8 +22,38 @@ func GetPubKeyHex(privateKey PrivateKey) string {
 	return hex.EncodeToString(privKey.PublicKey.X) + hex.EncodeToString(privKey.PublicKey.Y)
 }
 
-func GeneratePrivateKey() (PrivateKey, error) {
-	return elliptic.GeneratePrivateKey(elliptic.Secp256k1)
+func GeneratePrivateKey(difficulty int) PrivateKey {
+	for true {
+		privKey, err := elliptic.GeneratePrivateKey(elliptic.Secp256k1)
+		if err != nil {
+			fmt.Println("Error while generating private key", err)
```
Comment (Krigpl) [inline crypto/crypto.go]: remove this `Println` and make it return `err` as it used to

File: main/main.go [hunk 8]
```diff
@@ -66,9 +67,8 @@ func main() {
 	}
 
 	rand.Seed(time.Now().UTC().UnixNano())
-	privKey, err := crypto.GeneratePrivateKey()
-	if err != nil {
-		fmt.Println("Error while generating private key", err)
+	privKey := crypto.GeneratePrivateKey(KEY_DIFF)
```
Comment (Krigpl) [inline main/main.go]: the previous version with explicit error is better, revert to it

File: crypto/crypto.go [hunk 9]
```diff
@@ -20,8 +22,38 @@ func GetPubKeyHex(privateKey PrivateKey) string {
...
+			if pubKey[i] != 0 {
+				isDifficult = false
+				break
+			}
+		}
+
+		if isDifficult {
+			if pubKey[nullBytes] < (1 << (8 - remainder)) {
+				keyData := GetPubKeyHex(privKey)
+				ioutil.WriteFile("identity", []byte(keyData), 0600)
```
Comment (Krigpl) [inline crypto/crypto.go]: Move it outside this function, this function is supposed only to generate the key, not save it to storage as well.
Filename should also be in a constant.
Even better please make a separate PR with saving and reading (which is missing right now) the key to/from storage.

File: crypto/crypto.go [hunk 10]
```diff
@@ -20,8 +22,38 @@ func GetPubKeyHex(privateKey PrivateKey) string {
...
+
+		for i := 0; i < nullBytes; i++ {
+			if pubKey[i] != 0 {
+				isDifficult = false
+				break
+			}
+		}
+
+		if isDifficult {
+			if pubKey[nullBytes] < (1 << (8 - remainder)) {
```
Comment (Krigpl) [inline crypto/crypto.go]: Those two `if`s can be a single one

File: crypto/crypto.go [hunk 11]
```diff
@@ -20,8 +22,38 @@ func GetPubKeyHex(privateKey PrivateKey) string {
...
+		if err != nil {
+			fmt.Println("Error while generating private key", err)
+			return nil
+		}
+
+		xPart := make([]byte, len(privKey.PublicKey.X))
+		copy(xPart, privKey.PublicKey.X)
+		pubKey := sha256.Sum256(append(xPart, privKey.PublicKey.Y...))
+
+		nullBytes := difficulty / 8
```
Comment (Krigpl) [inline crypto/crypto.go]: Make this whole difficulty check in a separate function and write a test for it.

**Review by Krigpl — CHANGES_REQUESTED** [review]

File: crypto/crypto.go (hunk 7 above)
Comment (Elfoniok) [reply crypto/crypto.go]: Done

File: main/main.go (hunk 8 above)
Comment (Elfoniok) [reply main/main.go]: Done

File: crypto/crypto.go [hunk 12]
```diff
@@ -24,6 +24,43 @@ func GeneratePrivateKey() (PrivateKey, error) {
 	return elliptic.GeneratePrivateKey(elliptic.Secp256k1)
 }
 
+func GenerateDifficultKey(difficulty uint) (PrivateKey, error) {
+	for true {
```
Comment (Krigpl) [inline crypto/crypto.go]: nit: `for {` is the same

File: crypto/crypto.go [hunk 13]
```diff
@@ -24,6 +24,43 @@ func GeneratePrivateKey() (PrivateKey, error) {
...
+	copy(xPart, key.PublicKey.X)
+	pubKey := sha256.Sum256(append(xPart, key.PublicKey.Y...))
+
+	nullBytes := difficulty / 8
+	remainder := uint(difficulty % 8)
+	isDifficult := true
+
+	for i := uint(0); i < nullBytes; i++ {
+		if pubKey[i] != 0 {
+			isDifficult = false
```
Comment (Krigpl) [inline crypto/crypto.go]: just `return false` and get rid of `isDifficult` altogether

File: crypto/crypto.go [hunk 14]
```diff
@@ -24,6 +24,43 @@ func GeneratePrivateKey() (PrivateKey, error) {
...
+			break
+		}
+	}
+
+	if isDifficult && pubKey[nullBytes] < (1<<(8-remainder)) {
```
Comment (Krigpl) [inline crypto/crypto.go]: and then this becomes
`return pubKey[nullBytes] < (1<<(8-remainder))`

File: crypto/crypto.go [hunk 15]
```diff
@@ -24,6 +24,43 @@ func GeneratePrivateKey() (PrivateKey, error) {
 	return elliptic.GeneratePrivateKey(elliptic.Secp256k1)
 }
 
+func GenerateDifficultKey(difficulty uint) (PrivateKey, error) {
```
Comment (Krigpl) [inline crypto/crypto.go]: add a check and return an error if `difficulty > 256`

**Review by Krigpl — CHANGES_REQUESTED** [review]
just nits, otherwise lgtm

File: crypto/crypto.go [hunk 16]
```diff
@@ -25,7 +25,10 @@ func GeneratePrivateKey() (PrivateKey, error) {
 }
 
 func GenerateDifficultKey(difficulty uint) (PrivateKey, error) {
-	for true {
+	if difficulty > 256 {
+		return nil, errors.New("Difficulty to high")
```
Comment (Krigpl) [inline crypto/crypto.go]: `s/to/too`
Also the convention for error messages in Go is to start them with non capital letter in cases like this (so that later you can do `fmt.Println("An error occurred:", err)` and you won't get capital letter in the middle of the sentence

**Review by Krigpl — APPROVED** [review]
I don't know what happened but looks like you already merged part of this PR to master? Was that intended?
